### PR TITLE
Add 'break' command

### PR DESCRIPTION
**Pull request by perlun**
**Note**: this is not yet feature complete, it crashes when you press the `b` button:

```
$ GIT_SEQUENCE_EDITOR="./index.js" git rebase -i master
hint: Waiting for your editor to close the file... TypeError: Cannot read property 'forEach' of undefined
    at Terminal.render (/storage/git_3rd_party/rebase-editor/lib/terminal.js:104:16)
    at /storage/git_3rd_party/rebase-editor/lib/main.js:33:22
    at /storage/git_3rd_party/rebase-editor/lib/terminal.js:35:7
    at Terminal.listenerWrapper (/storage/git_3rd_party/rebase-editor/node_modules/nextgen-events/lib/NextGenEvents.js:282:36)
    at Function.emitToOneListener (/storage/git_3rd_party/rebase-editor/node_modules/nextgen-events/lib/NextGenEvents.js:500:36)
    at Function.emitEvent (/storage/git_3rd_party/rebase-editor/node_modules/nextgen-events/lib/NextGenEvents.js:421:17)
    at Terminal.emit (/storage/git_3rd_party/rebase-editor/node_modules/nextgen-events/lib/NextGenEvents.js:359:23)
    at Terminal.onStdin (/storage/git_3rd_party/rebase-editor/node_modules/terminal-kit/lib/Terminal.js:964:9)
    at ReadStream.emit (events.js:315:20)
    at addChunk (internal/streams/readable.js:309:12)
error: There was a problem with the editor './index.js'.
```

### CODE REVIEW

File: lib/reducer.js
Comment (perlun) [inline lib/reducer.js:31]: @sjurba I think this is the main problem with this PR; this method doesn't add the `break` line _and_ it messes up the state object so badly that the program crashes. :see_no_evil: Please advise if you have any ideas on how to make it work.

File: lib/reducer.js
Comment (real-yfprojects) [inline lib/reducer.js:201]:
```suggestion
    else if (action === 'break') {
      state = set(state, insertAfterCurrentPosition(state));
    }
```

File: lib/reducer.js
Comment (real-yfprojects) [inline lib/reducer.js:31]: I don't know javascript but that's my suggestion:

```suggestion
function insertAfterCurrentPosition(state, content) {
    let newLines = state.lines;
    let line = {action : 'break', hash : '', message : 'break'}
    insertInto(newLines, line,  state.cursor.pos + 1);
    return {
        lines: newLines,
        cursor: {
            from: state.cursor.from + 1,
            pos: state.cursor.pos + 1
        }
    }
}

```

**Review by real-yfprojects — CHANGES_REQUESTED** [review]

**Comment by perlun** [comment]
@real-yfprojects Not bad at all! :bow: Thanks for these changes. They weren't _entirely_right, but they brought me closer to a working version. With some minimal changes on top of your commits, I now have **working break support** (tested with `GIT_SEQUENCE_EDITOR="./index.js" git rebase -i master` locally).

@sjurba Should be ready for review now. :+1: :tada:

**Review by real-yfprojects — CHANGES_REQUESTED** [review]
I think one isn't able to "drop" a _break_ command.

**Comment by real-yfprojects** [comment]
> With some minimal changes on top of your commits, I now have **working break support**/

Great!
If I am not mistaken, you change the message to be blank instead of `break`.

File: lib/reducer.js
Comment (perlun) [reply lib/reducer.js:33]: Fixed, resolving.

**Comment by perlun** [comment]
> If I am not mistaken, you change the message to be blank instead of `break`.

Yeah. With the changes you suggested, it would appear as `break break` in the list of rebase commands, which made git complain that it wasn't a valid command (don't have the exact error message in front of me right now).

> I think one isn't able to "drop" a _break_ command.

Hmm, what do you mean with this? :thinking:

**Comment by real-yfprojects** [comment]
> Hmm, what do you mean with this? :thinking:

A way to remove an added break action. I think this should be possible by hitting `d` when a line with a break is selected.

**Comment by perlun** [comment]
> A way to remove an added break action. I think this should be possible by hitting d when a line with a break is selected.

Hmm, you're right. It doesn't work at the moment. :thinking: I tried fixing it with the following diff:

```diff
diff --git lib/reducer.js lib/reducer.js
index 90c457b..aa23493 100644
--- lib/reducer.js
+++ lib/reducer.js
@@ -81,7 +81,7 @@ function getAction(state, action) {
   const [from, to] = getSelection(state);
   return {
     lines: state.lines.map((line, idx) => {
-      if (from <= idx && idx <= to && line.hash) {
+      if (from <= idx && idx <= to) {
         line = set(line, {
           action: action
         });
@@ -194,11 +194,10 @@ module.exports = function reducer(state, action, param) {
       });
     } else if (actions.includes(action)) {
       const [from, to] = getSelection(state);
-      if (state.lines.slice(from, to + 1).some((line) => line.action !== action && line.hash)) {
+      if (state.lines.slice(from, to + 1).some((line) => line.action !== action)) {
         state = set(state, getAction(state, action));
       }
-    }
-    else if (action === 'break') {
+    } else if (action === 'break') {
       state = set(state, insertAfterCurrentPosition(state));
     }
     if (oldState !== state && oldState.lines !== state.lines) {
```

(i.e. removing the `hash` check)

However, this breaks when you try to use it (since `drop` doesn't work on its own like that):

```
$ GIT_SEQUENCE_EDITOR="./index.js" git rebase -i master
error: missing arguments for drop
error: invalid line 4: drop
error: missing arguments for drop
error: invalid line 5: drop
error: missing arguments for drop
error: invalid line 6: drop
error: missing arguments for drop
error: invalid line 7: drop
error: missing arguments for drop
error: invalid line 8: drop
error: missing arguments for drop
error: invalid line 9: drop
You can fix this with 'git rebase --edit-todo' and then run 'git rebase --continue'.
Or you can abort the rebase with 'git rebase --abort'.
```

I think the "proper" fix is to special-case `drop` in case `break` was used and actually remove the line instead. This feels like a bit more work though and I would be willing to say "later" (after this PR is merged) for that. :slightly_smiling_face: If @sjurba has any good ideas, let me know.

**Comment by perlun** [comment]
Thanks @real-yfprojects. I tried this approach but it doesn't work; `drop` or the Delete key doesn't do anything. :cry: Could you give it a try locally? (just take my branch + the change you have above and run `GIT_SEQUENCE_EDITOR="./index.js" git rebase -i master`)

**Comment by real-yfprojects** [comment]
I couldn't run it locally because it misses the terminal-kit module. But I fixed it any ways in my PR.

**Comment by perlun** [comment]
@sjurba There we go, with a lot of great help from @real-yfprojects (thanks! :bow:) I think we have a version now that should be ready to be reviewed and potentially merged into your tree.

**Review by real-yfprojects — APPROVED** [review]

**Comment by sjurba** [comment]
Awesome work guys. Did you by the way check if the tests are still working? And perhaps I could bother you to write a test or two for the new functionality as well? As I remember I had quite a good test coverage for the code, so it would be great to maintain that..

**Comment by perlun** [comment]
> Awesome work guys. Did you by the way check if the tests are still working?

Hmm, good point. :thinking: `yarn test` gives me a few test failures in fact:

```
  115 passing (166ms)
  3 failing

  1) Debug log should write console log to file:

      AssertionError: expected 'Jalla\n  \u001b[32m  ✓\u001b[0m\u001b[90m should trap console log\u001b[0m\n' to equal 'Jalla\n'
      + expected - actual

       Jalla
      -    ✓ should trap console log
      
      at test/debug-log.spec.js:31:58

  2) Reducer change action key should do nothing if action is not different:

      AssertionError: expected { Object (lines, cursor, ...) } to equal { Object (lines, cursor, ...) }
      + expected - actual

         ]
         "otherStateVar": {
           "foo": "bar"
         }
      -  "redoStack": []
      -  "undoStack": [
      -    {
      -      "cursor": {
      -        "from": 0
      -        "pos": 0
      -      }
      -      "lines": [
      -        {
      -          "action": "pick"
      -          "hash": "123"
      -          "message": "Line 0"
      -        }
      -      ]
      -    }
      -  ]
       }
      
      at Context.<anonymous> (test/reducer.spec.js:325:29)
      at processImmediate (internal/timers.js:461:21)

  3) Reducer change action key should do nothing if line has no hash:

      AssertionError: expected { Object (lines, cursor, ...) } to equal { Object (lines, cursor, ...) }
      + expected - actual

         ]
         "otherStateVar": {
           "foo": "bar"
         }
      -  "redoStack": []
      -  "undoStack": [
      -    {
      -      "cursor": {
      -        "from": 0
      -        "pos": 0
      -      }
      -      "lines": [
      -        {
      -          "action": "noop"
      -        }
      -      ]
      -    }
      -  ]
       }
      
      at Context.<anonymous> (test/reducer.spec.js:333:29)
      at processImmediate (internal/timers.js:461:21)



error Command failed with exit code 3.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

> And perhaps I could bother you to write a test or two for the new functionality as well? As I remember I had quite a good test coverage for the code, so it would be great to maintain that..

Would clearly not hurt. I've done "manual testing" with `GIT_SEQUENCE_EDITOR="./index.js" git rebase -i master` for now.

@real-yfprojects - if you like, you could give these things a look? I don't think I'll have time for it myself. (disclosure: I've been doing this on "work-time" until now but going deeper into the details on how this project works feels perhaps more like something I wouldn't be comfortable doing at my employer's expense.)

**Comment by real-yfprojects** [comment]
> @real-yfprojects - if you like, you could give these things a look? I don't think I'll have time for it myself. (disclosure: I've been doing this on "work-time" until now but going deeper into the details on how this project works feels perhaps more like something I wouldn't be comfortable doing at my employer's expense.)

I have been doing this in my free time. Since I wouldn't consider myself a javascript developer and wouldn't claim to know javascript either I coded the changes in this PR mostly from looking at code that was already in place. Currently I don't have the time to learn javascript or dive deeper into the test framework and such which will probably need further understanding of the javascript language.

**Comment by real-yfprojects** [comment]
> Break is now supported! partying_face

Great!

**Comment by real-yfprojects** [comment]
Still I am a little disappointed that large parts of this PRs work didn't find their way into the master branch. Almost all the issues linked to the behaviour of the cursor or selection Perlun found in perlun#1 now apply to the master branch although they were fixed in perlun#1.

**Comment by sjurba** [comment]
Ah. Right, I didn't notice the issue with the cursor. I'll get that fixed! Where there more issues you where concerned about..?

**Comment by real-yfprojects** [comment]
- [ ] Cursor out of bound when deleting last line
- [ ] Cursor moves up when deleting a single line
- [ ] Selection isn't adjusted when deleting lines
- [ ] When hitting `b` when there is a break statement in the following line the cursor doesn't move to that line

**Comment by perlun** [comment]
> Still I am a little disappointed that large parts of this PRs work didn't find their way into the master branch. Almost all the issues linked to the behaviour of the cursor or selection Perlun found in [perlun#1](https://github.com/perlun/rebase-editor/pull/1) now apply to the master branch although they were fixed in [perlun#1](https://github.com/perlun/rebase-editor/pull/1).

Yeah, it's not the end of the world for me, but it does feel a bit silly that I did actually spend a bit of time (iterating over this with @real-yfprojects), testing and testing it over again (manually with `GIT_SEQUENCE_EDITOR="./index.js" git rebase -i master`), only to hear that these bugs that we fixed ended up making their way into the `master` branch anyway. :joy:

I don't want to be too negative, I am really happy about the time and effort you've put into this project @sjurba and I'm not even sure I would have handled this better myself. Sometimes it's just easier to "fix issues" as a maintainer rather than spending a lot of time trying to explain the inner workings of the test suite and what not.

It seems a bit like this illustrates also why manual testing sometimes has an advantage over automated testing (and no, I'm of course not against that per se... I spend a lot of time writing integration tests in other projects). It's just that spending all the effort into making the automated tests cover _everything_ can be quite a time-consuming process.

----

Anyway, I'm happy that this is moving forward and looking forward to seeing these bugs fixed in `master` eventually. Until then, I'll keep using my local branch instead. :wink:

**Comment by sjurba** [comment]
So I fixed the issues you listed with the exception of moving up when deleting a line. I kept it because it simplifies the code and I find that it is the most natural thing to do as it works the same as if you press the erase button on the keyboard. Moving it down would be more like pressing the delete key I guess. Anyways, I don't see that as a big issue.

I'm sorry that you are left feeling disappointed and I really do appreciate the efforts you have made. The reason I ended up rewriting parts of the code was that your implementation actually broke some of the tests, perfectly illustrating why I spent so much time writing them and also why I feel it is important to keep them updated. You had removed some lines that ended up breaking some code I had for dealing with #noop lines I think. The code base is quite intricate with lots of corner cases that I discovered along the way through manual testing like you did. The only way to make sure I can refactor or change the code while making sure I don't break some corner cases I have long forgotten about is to write tests for them. 

Also having a test suite that makes it easy to write new tests, actually makes it a lot faster to develop the code as well. Constantly going back and and forth and testing manually takes much longer than having some tests running and verifying your code in less than a second. Of course it does not remove the need for manual testing. You have to test and run the code manually to find and check for corner cases and to make sure the code actually behaves in the way you want. But once you identified the corner cases, writing a test first before trying to fix it actually is much faster that with just doing it manually. 

Now when it comes to 100% coverage, I do admit it's not really that useful. I think about 95% of the tests are actually useful. The rest are just there because I got so close I might as well do the rest.

**Comment by perlun** [comment]
Thanks for the comments @sjurba. Nice that you got (most of) these bugs sorted out. :+1: 

I'm totally with you on the "keep tests working" and "keep the code covered" with automatic tests. As I mentioned, that's the way I work in other projects. It's definitely the way to make sure you don't introduce regressions into the codebase at a later time. In this particular case, I guess I've been more sloppy, and I appreciate you taking the time to ensuring that the tests are working and that the new functionality is properly (unit) tested as well.

I tested your version now. It's quite fine, even though there are semantic differences to the version me and @real-yfprojects worked on. For example, in "our version", it was possible to add multiple `break` lines - probably not an important "feature" in real world so perhaps just as good that you dropped it.

As for this part:

> So I fixed the issues you listed with the exception of moving up when deleting a line. I kept it because it simplifies the code and I find that it is the most natural thing to do as it works the same as if you press the erase button on the keyboard. Moving it down would be more like pressing the delete key I guess. Anyways, I don't see that as a big issue.

...I'm not sure I agree actually. It feels odd to let the _Delete_ key (which also can be used to remove a line in addition to `d`) work like this. I really think you should consider making it work like the "Delete" button normally works, i.e "forward-delete" and not "backward-delete". In this case, I think the UX and end-user experience of the functionality would be more important than simplifying the code in fact. (note: I'm writing this as an innocent bystander, since I don't really know how much more complex it'll make it)

(Wishing you both a nice Easter holiday, I'll probably be a bit slower to respond during the next few days. :pray:)

**Comment by real-yfprojects** [comment]
> ...I'm not sure I agree actually. It feels odd to let the _Delete_ key (which also can be used to remove a line in addition to `d`) work like this. I really think you should consider making it work like the "Delete" button normally works, i.e "forward-delete" and not "backward-delete". In this case, I think the UX and end-user experience of the functionality would be more important than simplifying the code in fact. (note: I'm writing this as an innocent bystander, since I don't really know how much more complex it'll make it)

I would also agree that UX is more important than code complexity but I find this UX improvement not that important.

**Comment by real-yfprojects** [comment]
Nice holidays to you as well!

**Comment by perlun** [comment]
> So I fixed the issues you listed with the exception of moving up when deleting a line. I kept it because it simplifies the code and I find that it is the most natural thing to do as it works the same as if you press the erase button on the keyboard. Moving it down would be more like pressing the delete key I guess. Anyways, I don't see that as a big issue.

@sjurba I still think Delete-like semantics would be preferable in this case. Would you be willing to accept a PR that fixes it like this, if it doesn't complicate the code too much and still keeps all the tests passing?

**Comment by sjurba** [comment]
So the discussion is really Delete vs. Backspace. The drop command is bound to both DELETE and BACSPACE, for normal drop it doesn't matter as the cursor does not move. Only when introducing the break keyword did we start moving the cursor and this became an issue. No matter which way we chose, either DELETE or BACKSPACE will be "wrong". And since both solutions have the same downside, I chose the BACKSPACE solution since then you don't have to consider wether you are on the last line or not. You don't have the symetrical problem of being on the first line as you can never (makes no sense to) start with a break on the first line. 

The only no-compromise solution is to have separate actions for DELETE and BACKSPACE, which is of course possible, but a bit of overkill? How often do you add a break and then delete it again? I can't really see how this is a big enough deal to spend time on. 

On the other side, it would be great to tackle the exec and label, merge stuff at some point, at which time figuring out how delete should work will become a (bigger) issue again. In that case I would be more inclined to supporting both delete and backspace properly.. 

With this in mind would you still like to spend time on this issue? I could be inclined to merge a nice solution separating DELETE and BACKSPACE into separate commands, assuming all tests pass with 100% coverage.

**Comment by perlun** [comment]
> The only no-compromise solution is to have separate actions for DELETE and BACKSPACE, which is of course possible, but a bit of overkill? How often do you add a break and then delete it again? I can't really see how this is a big enough deal to spend time on.

Yeah, maybe you're right. I think the main (hypothetical) use case would be that you add a `break` line in the wrong place, realizing it, pressing Delete on it and then the cursor moving in an unexpected way. But yes, this could clearly be the "edge case of edge cases", it's perhaps more likely to happen when testing the feature manually than in real-world usage. :slightly_smiling_face: 

Worth mentioning: I always tend to use the Delete button in `rebase-editor` for some reason. I never seem to be using Backspace.

> On the other side, it would be great to tackle the exec and label, merge stuff at some point, at which time figuring out how delete should work will become a (bigger) issue again. In that case I would be more inclined to supporting both delete and backspace properly..

Yeah, I think this sounds good. :+1: I agree that making a distinction between Delete and Backspace seems like the Way To Go in this case.

Given that you are the "main brain" behind this and you know how to implement this well, and that the code wasn't immediately easy to understand (for me) when I briefly looked at it... I think I'll just be lazy in this case and wait for you to fix it "at some point" (with 100% coverage). :grin: I am actually even using my old branch right now, which has the "Delete-style" semantics. :see_no_evil:

**Pull request closed without merging**
